services/search/src/search/routes/search.py

```diff
@@ -46,33 +47,74 @@
 
 from search.duckdb_connection import duckdb_connect
 
-FTS_COMMAND_COUNT = (
-    "SELECT COUNT(*) FROM (SELECT __hf_index_id, fts_main_data.match_bm25(__hf_index_id, ?) AS __hf_fts_score FROM"
-    " data) A WHERE __hf_fts_score IS NOT NULL;"
-)
+FTS_COUNT_COMMAND = "SELECT COUNT(*) FROM (SELECT __hf_index_id, fts_main_data.match_bm25(__hf_index_id, ?) AS __hf_fts_score FROM data) A WHERE __hf_fts_score IS NOT NULL;"
+
 
-FTS_COMMAND = (
+FTS_BY_TABLE_COMMAND = (
     "SELECT * EXCLUDE (__hf_fts_score) FROM (SELECT *, fts_main_data.match_bm25(__hf_index_id, ?) AS __hf_fts_score"
     " FROM data) A WHERE __hf_fts_score IS NOT NULL ORDER BY __hf_fts_score DESC OFFSET {offset} LIMIT {length};"
 )
 
+FTS_BY_BATCHES_COMMAND = "SELECT * FROM (SELECT *, fts_main_data.match_bm25(__hf_index_id, ?) AS __hf_fts_score FROM data WHERE __hf_index_id BETWEEN {offset} and {length}) A WHERE __hf_fts_score IS NOT NULL;"
+
+FTS_BATCH_SIZE = 1000
 
 logger = logging.getLogger(__name__)
 
+DATASETS_WITH_FTS_BY_BATCHES = ["wikimedia/wikipedia", "asoria/bolivian-recipes"]
+
+
+def _search_by_table(con: DuckDBPyConnection, query: str, offset: int, length: int) -> pa.Table:
+    query_result = con.execute(
+        query=FTS_BY_TABLE_COMMAND.format(offset=offset, length=length),
+        parameters=[query],
+    )
+    return query_result.arrow()
+
+
+def _search_by_batches(
+    num_original_rows: int, con: DuckDBPyConnection, query: str, offset: int, length: int
+) -> pa.Table:
+    num_matched_rows = 0
+    index = 0
+    pa_tables = []
+
+    for index in range(index, num_original_rows, FTS_BATCH_SIZE):
+        fts_query = FTS_BY_BATCHES_COMMAND.format(offset=index, length=index + FTS_BATCH_SIZE)
+        logging.info(fts_query)
+        pa_table = con.execute(
+            query=fts_query,
+            parameters=[query],
+        ).arrow()
+        pa_tables.append(pa_table)
+        num_matched_rows += pa_table.num_rows
+        logging.info(f"{pa_table.num_rows} rows for batch {index} - {num_matched_rows} for table")
+        if num_matched_rows < (offset + length):
+            break
+    pa_result = pa.concat_tables(pa_tables)
+    pa_result = pa_result.sort_by([("__hf_fts_score", "descending")]).drop("__hf_fts_score")
```

Comment:
weren't we returning the results by row_idx, instead of FTS score?

Reply:
I don't quite remember the related issue, but It was done in this PR https://github.com/huggingface/datasets-server/pull/1790